Library/src/Multiplayer/EventBus.cpp

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2023 Magnopus LLC
+
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "CSP/Multiplayer/EventBus.h"
+
+#include "CSP/Systems/SystemBase.h"
+#include "CallHelpers.h"
+#include "Debug/Logging.h"
+#include "Memory/Memory.h"
+#include "Multiplayer/SignalR/SignalRConnection.h"
+#include "NetworkEventManagerImpl.h"
+
+namespace csp::multiplayer
+{
+
+extern ErrorCode ParseError(std::exception_ptr Exception);
+
+constexpr const uint64_t ALL_CLIENTS_ID = -1;
+
+EventBus::~EventBus()
+{
+	CSP_DELETE(MultiplayerConnectionInst);
+}
+
+EventBus::EventBus(const EventBus& InBoundConnection)
+{
+	MultiplayerConnectionInst = InBoundConnection.MultiplayerConnectionInst;
+	EventMap				  = InBoundConnection.EventMap;
+}
+
+EventBus::EventBus(MultiplayerConnection* InMultiplayerConnection)
+{
+	MultiplayerConnectionInst = InMultiplayerConnection;
+	EventMap				  = {};
+}
+
+void EventBus::ListenEvent(const csp::common::String& EventName, ParameterisedCallbackHandler Callback, csp::systems::SystemBase* System)
+{
+	if (MultiplayerConnectionInst->Connection == nullptr || !MultiplayerConnectionInst->Connected)
+	{
+		return;
+	}
+
+	if (!Callback && !System)
+		return;
+
+	if (Callback)
+	{
+		EventMap[EventName].first.push_back(Callback);
+	}
+	if (System)
+	{
+		EventMap[EventName].second = System;
+	}
+}
+
+void EventBus::ListenEvent(const csp::common::String& EventName, csp::systems::SystemBase* System)
+{
+	ListenEvent(EventName, nullptr, System);
+}
+
+void EventBus::ListenEvent(const csp::common::String& EventName, ParameterisedCallbackHandler Callback)
+{
+	ListenEvent(EventName, Callback, nullptr);
+}
+
+void EventBus::StopListenEvent(const csp::common::String& EventName)
+{
+	if (!EventMap.empty() && EventMap.find(EventName) != EventMap.end())
+	{
+		EventMap.erase(EventName);
+	}
+}
+
+void EventBus::StartEventMessageListening()
+{
+	std::function<void(signalr::value)> LocalCallback = [this](signalr::value Result)
+	{
+		if (Result.is_null())
+		{
+			return;
+		}
+
+		if (EventMap.empty())
+		{
+			return;
+		}
+
+		std::vector<signalr::value> EventValues = Result.as_array()[0].as_array();
+		const csp::common::String EventType(EventValues[0].as_string().c_str());
+
+		if (EventMap.find(EventType) == EventMap.end())
+		{
+			return;
+		}
+
+		if (EventMap[EventType].second)
+		{
+			EventMap[EventType].second->Deserialise(EventValues);
+		}
+		else
```

Comment:
This means that if a system has been registered any registered callbacks will be ignored, is that correct? If that is by design then I would not have a ListenEvent method that takes both a callback and an event.

Reply:
that is a good point. I was not sure if this would ever be a use-case, so I kind of left it half-baked. apologies! so is there a use case where the same event would be listened to both by a System and by one or more ParameterisedCallbackHandlers?
If not, would it make sense to decouple the two completely and have 2 separate event maps? 
If yes, I need to remove the else here and see if other changes are required

Reply:
Hmmm. Thihking about it the whole point of being able to register systems is because they have unique deserialisation requirements. In which case there would be no point sending those events to a system to be deserialised and to then apply the default deserialiser for any registered callbacks.

So I think it should be that you either register one or many callbacks for an event (which then use the default deserialiser), or you register a single system (which then uses its unique deserialiser). You don't have both.

If we are doing that then we should ensure that people do not register a system and calback/s for the same event.

What do you think @MAG-mv

Reply:
Yep. That sounds good